Reliable Mutability Primitive pitch (#19)
protocol/web3-dev-team: proposals/Reliable Mutability Primitive.md

> <!--

Author: ```suggestion
# Reliable Mutability Primitive

Authors: [@gozala](https://github.com/gozala)

Initial PR: https://github.com/protocol/web3-dev-team/pull/19/
<!--
```

This template is intended to be used by those who would like to pitch a new project for one of the Web3 Dev project teams to take on. It should contain sufficient detail that others can understand how this project contributes to our team’s mission of  product-market fit for our unified stack of protocols, what is included in scope of the project, where to get started if a project team were to take this on, and any other information relevant for prioritizing this project against others.
Good project scope aims for ~3-5 engineers for 1-3 months (though feel free to suggest larger-scoped projects anyway). Projects do not include regular day-to-day maintenance and improvement work, e.g. on testing, tooling, validation, code clarity, refactors for future capability, etc.
-->

> # Reliable Mutability Primitive

Author: ```suggestion
```

### Purpose &amp; impact 
##### Background &amp; intent
_Describe the desired state of the world after this project? Why does that matter?_
<!--
Outline the status quo, including any relevant context on the problem you’re seeing that this project should solve. Wherever possible, include pains or problems that you’ve seen users experience to help motivate why solving this problem works towards top-line objectives. 
-->

Today ecosystem has to work around the lack of functional mutability primitives. [IPNS][] does not provide reliable solution forcing community to adopt different custom solutions.

> - [Pinata][] gets asked about how to chaneg content for CID onec every few weeks _(Most of the time they find that is not what the users actually need)_.

Reviewer: minor typos - `change`, `once`

- [fission.codes](https://fission.codes/) found IPNS persistance unreliable and had to build custom solution that updates [DNSLink][] records (**TODO**: Get a quote / link from Boris)
- **TODO**: Survey textile, last time we talked they found IPNS to be unreliable for propagating thread updates.
- ...
- Protocol Labs uses dns simple API to [publish site updates](https://github.com/filecoin-project/specs/blob/71f37208a1f4f56b33ea307d7cbdb4b06996b115/.github/workflows/main.yml#L40). 

> This gap in the stack makes building software which runs on end user computers and meets modern expectations more difficult than traditional hosted solutions. This creates a hardle for adoption of the technology.

Reviewer 2: ```suggestion
This gap in the stack makes building software which runs on end user computers and meets modern expectations more difficult than traditional hosted solutions. This creates a hurdle for adoption of the technology.
```

##### Assumptions &amp; hypotheses
_What must be true for this project to matter?_
<!--(bullet list)-->

- Applications want to provide a canonical address for a current state.
- Application authors will prefer built-in mutability primitive over custom solutions. 
- Developed solution would provide simpler mechanism for publishing updates (than DNSLink updates)

> - Developed solution would work provide reliable mechanism for producing updates on not always connected end user computers (think browsers, phones)

Reviewer 2: ```suggestion
- Developed solution would provide a reliable mechanism for producing updates on not always connected end user computers (think browsers, phones)
```

# Reliable Mutability Primitive

Authors: [@gozala](https://github.com/gozala)

> Initial PR: https://github.com/protocol/web3-dev-team/pull/19/

Reviewer 3: Maybe add a note that this is part of the larger story in https://github.com/protocol/web3-dev-team/pull/42 ?

Author: I am not sure this would be accurate. I think there is a merit to reliable mutability primitive whether we decide to do human readable names or not.

-->

### Purpose &amp; impact 
##### Background &amp; intent
_Describe the desired state of the world after this project? Why does that matter?_
<!--
Outline the status quo, including any relevant context on the problem you’re seeing that this project should solve. Wherever possible, include pains or problems that you’ve seen users experience to help motivate why solving this problem works towards top-line objectives. 
-->

Today ecosystem has to work around the lack of functional mutability primitives. [IPNS][] does not provide reliable solution forcing community to adopt different custom solutions.

- [Pinata][] gets asked about how to change content for CID once every few weeks _(Most of the time they find that is not what the users actually need)_.
- [fission.codes](https://fission.codes/) found IPNS persistance unreliable and had to build custom solution that updates [DNSLink][] records (**TODO**: Get a quote / link from Boris)

> - **TODO**: Survey textile, last time we talked they found IPNS to be unreliable for propagating thread updates.

Author: [@carsonfarmer: This is still correct, IPNS is too unreliable for anything with essentially sub-minute temporal resolution. So for data updates, it is not useful. However, we are finding it quite useful for buckets, and have used it to great effect recently.](https://hackmd.io/@gozala/ipns-pitch)

Reviewer 4: I should also mention something else to provide further context: We're currently managing 1000s (yes, that's right) of IPNS updates for buckets. This currently takes an extremely long time to republish. We are actively conducting tests to determine exactly how long this takes, and how often it fails to complete. In some cases, it is taking hours per day.

_How much would nailing this project improve our knowledge and ability to execute future projects?_

<!-- Explain the opportunity or leverage point for our subsequent velocity/impact (e.g. by speeding up development, enabling more contributors, etc)
-->

🎯🎯 - This would enable projects that need to have a canonical identifier for evolving dataset. It would also provide a greater interop across products built around IPFS by removing need for custom incompatibile solutions that teams need to implemente today.

##### Confidence
_How sure are we that this impact would be realized? Label from [this scale](https://medium.com/@nimay/inside-product-introduction-to-feature-priority-using-ice-impact-confidence-ease-and-gist-5180434e5b15)_.

<!--Explain why this rating-->

3 - As high confidence as we can have without doing actual user studies. We know teams that have tried IPNS but found it unrelaible, ultimately rolling out an alternative solution. We also know teams that have evaluated IPFS but chose alternative due to lack of reliable mutability story.

### Project definition
##### Brief plan of attack

<!--Briefly describe the milestones/steps/work needed for this project-->

> - Implement third party IPNS republishing (TODO: Find a relevant issue link)

Reviewer 5: This is a closely related issue https://github.com/ipfs/go-ipfs/issues/4435

> - Implement third party IPNS republishing (TODO: Find a relevant issue link)

Author: @carsonfarmer: [Yeh, this one shouldn't be ignored... maybe the existing solution is just not going to cut it in the end, and a fresh start is required. How likely is it that we could see "two" official methods, and have one of them win out?](https://hackmd.io/@gozala/ipns-pitch)

> - Implement third party IPNS republishing (TODO: Find a relevant issue link)

Author: ```suggestion
- Implement third party IPNS republishing (see [ipfs/go-ipfs#4435](https://github.com/ipfs/go-ipfs/issues/4435))
```

- Collaborate with Pinning Services to make this part of the API.

> - Fix IPNS publishing in browsers.

Reviewer 3: (New person comment): is this essential for unblock the core use-case of enabling customers to do ipns updates instead of DNS updates?

Author: Yes if those updates were to made from browser. It is also however useful on it's own merit as it enables apps to pull updated content even if such content has no human readable name.

> - Enable pubsub by default + IPNS over pubsub.

Reviewer 5: Note: Textile threads utilize pubsub which means that any issues with enabling pubsub by default are things that would be experienced by Textile threads users. They also utilize IPNS over PubSub for buckets.

> - Implement & deploy strategic republishing system on PL hosted nodes that will republish popular content and drop stalled one. So that IPNS works even thought pinning services.

Reviewer 5: Interesting, I'm not sure this is really required as we can give IPNS the same guarantees as IPFS. However, since the amount of data is smaller this idea of a public IPNS pinning service seems pretty cool.

> - (Nice to have) Implement and rollout delegated publishing to enable updating from multiple devices.

Reviewer 5: This is sort of doable already "for free" if we have some way to pin/follow IPNS updates. I can just share the private key with my multiple devices via `ipfs key export/import`.

There are alternative approaches too, but figured I'd point this one out.

Author: Yeah I meant getting it to a state where we can comfortably tell users this is how you go about it. I do have some reservations towards key sharing, but that's probably discussion worth having when we get here.

- (Nice to have) Implement and rollout versioning system so that nodes can see full update history.

##### What does done look like?
_What specific deliverables should completed to consider this project done?_

Publishing IPNS record takes just a few miliseconds. Shutting laptop and loading `https://ipfs.io/ipns/{key}` week later on other device loads the content published as fast as loading corresponding `https://ipfs.io/ipfs/{cid}`.

#####  What does success look like?
_Success means impact. How will we know we did the right thing?_

<!--
Provide success criteria. These might include particular metrics, desired changes in the types of bug reports being filed, desired changes in qualitative user feedback (measured via surveys, etc), etc.
-->

> People and teams are storing IPNS keys in [DNSLink][] records as opposed to CIDs, because that provides a more effective and simpler way to publish updates. We see new projects are leveraging IPNS over working around it via things like [DNSLink][].

Reviewer 5: At the very least because updating DNS is such a pain we see people using DNSLink and pointing it at IPNS records instead of IPFS because its so much easier to work with IPNS updates than DNS updates. Similarly, we may see people storing IPNS addresses in ENS to save on not wanting to publish many blockchain transactions.

> People and teams are storing IPNS keys in [DNSLink][] records as opposed to CIDs, because that provides a more effective and simpler way to publish updates. We see new projects are leveraging IPNS over working around it via things like [DNSLink][].

Author: ```suggestion
- People and teams are storing IPNS keys in [DNSLink][] records as opposed to CIDs, because that provides a more effective and simpler way to publish updates.
- We see people storing IPNS addresses in ENS to save on blockchain transactions costs.
- We see new projects leveraging IPNS (when human readable name is not a concern) instead of working around it wit [DNSLink][].
```

##### Counterpoints &amp; pre-mortem
_Why might this project be lower impact than expected? How could this project fail to complete, or fail to be successful?_

> - IPNS has bad reputation, changing it may prove to be a challenge.

Reviewer 3: How do we get more data on this?  Is this a question to PM's?

Author: I think it's well know fact, evidence been that teams have tried and ultimately end up going with DNSLink solution. I'm pretty sure PL did that too, but someone with longer history here might know better. 

We could also get specific collabs to comment here to either validate the claim or dispute it. I have invited folks from textile & fission to validate. We could reach out to more teams.

@olizilla do you by a chance know if we used to have IPNS in DNSLink past or am I making wrong assumptions here.

Reviewer 6: @BigLep good pulse check in this issue (2017–today): https://github.com/ipfs/go-ipfs/issues/3860

> - Teams may find that general primitive is not adequate for custom needs.

Reviewer 3: Have we been presented any usecases where ipns (once it has a reliable mutability primitive) isn't sufficient?

Author: - IPNS lacks human readable/memorable naming which is a concern in most human facing interactions. 
- I have in the past voiced concerns about the lack of revision history associated with it (although one could built such a system on top).
- There also had been discussions about ability to have multiple publishers to the same name.
- And then there are also questions about compromised or lost key recovery

I suspect that despite all this usable IPNS will address enough use cases to be worth it. That said we might find that a higher level primitive that address more use cases out of the box will provide a better market fit.

- Required key management (e.g. key recovery, key compromise) still may prove it impractical.
- Higher level alternatives like textile threads may prove to be better layer of abstraction.

##### Alternatives
_How might this project’s intent be realized in other ways (other than this project proposal)? What other potential solutions can address the same need?_

> - Integrating textile threads as core component of the ecosystem.

Author: @carsonfarmer: [Obviously I like this idea. But it would end up being something threads-like I think. A benefit here is that threads supports multi-writer updates, which I think its a feature lacking in IPNS right now. The downside is that threads "as is" won't scale. Additionally, the encryption layer in threads is too heavy handed right now for a general solution like this. Having said all that, these things are all fixable implementation details.](https://hackmd.io/@gozala/ipns-pitch)

> - Standardizing DNS record updates mechanism and intergating into IPFS.

Reviewer 3: Is this even possible?  Doesn't it depend on each DNS' API for mutating actions?

Author: Hypothetically speaking we can start a standardization effort form a committee  and get enough big players onboard to get it done. I imagine there might be some value for DNS providers to make their products more accessible.

> - Work with Pinning Services to implement stardized DNSLink updating mechanism.

Author: @carsonfarmer: [This would be an interesting shorter term solution](https://hackmd.io/@gozala/ipns-pitch)

Reviewer 3: (new person coming up to speed): what would this look like?  I think I'll be able to grok it with a tiny sketch or a mock sample code example.

Author: We have recently created [pinning service API](https://ipfs.github.io/pinning-services-api-spec/#section/Schemas) that allows you to add pinning services like pinata to your IPFS node and then remote pin things. Kind of like adding git remote and pushing changes to it.

Idea here is to extend that API in a way that you could update DNSLink provided by the pinning service to the new CID through extended API. Kind of like when you push to remote it updates corresponding branch to new commit. In our case branch names would be DNSLinks, that you could update to point to new CID.

@BigLep does this answer your question ?

Reviewer 3: @Gozala : thanks - this helps.  The only open area for me then is where is this DNSLink stored?  Is this a DNSLink in a DNS record on a (sub)domain that the Pinning Service owns?  

Basically would the chain be
``My domain (e.g., mywebsite.com) has DNS record that points to a domain the Pinning service controls (e.g., mywebsite.somepinningservice.com) which has the DNSLink to the latest CID?``

Reviewer 6: @BigLep Depends if you (A) publish to your own domain, or  (B) use subdomain provided by pinning service for free. 

To give you a real life example, let's say  we want to host https://cid.ipfs.io on some pinning service (pin + take care of updating dnslink + TLS).

(B) is easy. Pinning service simply give you a name on some domain they own. Fleek does that already, they give  https://cid-utils-website.on.fleek.co/  for free + they set up DNSLink for those names (because they own the top domain).

(A) means we want to delegate updates of  DNS TXT record on  `_dnslink.cid.ipfs.io` to some pinning service. 
This can be done either by:
- giving DNS provider credentials to pinning service + expecting them to implement support for proprietary APIs of each dns provider (not realistic IMO, but mentioning here for completeness)
-  using DNS protocol itself for delegation: setting NS records that delegate management of `_dnslink.cid.ipfs.io` to a DNS server controlled by pinning service. That way they have full control over DNSLink without having any control over other records (because we delegated only `_dnslink` subdomain.

##### Dependencies/prerequisites
<!--List any other projects that are dependencies/prerequisites for this project that is being pitched.-->

> - Functioning IPNS in browsers

Reviewer 5: Note: likely includes an up-to-date IPNS over PubSub implementation

Author: @aschmahmann do you mean something like 

```suggestion
- Functioning IPNS in browsers (over up-to-date PubSub implementation)
```

Or a separate item ? I

Reviewer 5: Well IPNS over PubSub has a few tweaks over just PubSub itself to give some reliability. It's specced [here](https://github.com/ipfs/specs/blob/a370a6aa2345b251a4c04d37b8c50d3aa8f2701f/naming/pubsub.md).

@hugomrdias would probably know more about other work required to get IPNS functioning in a browser I just know that without the IPNS over PubSub changes in the spec that it won't be fast enough.

Reviewer 3: Sorry for the basic question, but I'm still confused by what is meant by "functioning IPNS in browsers".  What isn't working browsers today?

Author: I should probably capture this in the document. Currently pubsub is disabled in go-ipfs so all the IPNS records are published and resolved from DHT. Nodes in browsers do not have necessary capabilities to query/publish DHT. Instead they do publish and resolve IPNS records over pubsub. Result is fragmented network where go and js nodes see each others records.

Furthermore because because there is not enough nodes on the network with pubsub enabled published records from browser don't always make to nodes subscribed to them.

P.S.: Some of this may not be completely accurate, @hugomrdias @aschmahmann could you please confirm I'm not misrepresenting current state.

- Enabling pubsub in go-ipfs

##### Future opportunities
<!--What future projects/opportunities could this project enable?-->

- [Petname][] system for memorable but non canonical names (kind of how our phones allow us to not remember numbers).
- Subscribtion based deny lists (node can subscribe to any endpoint)
- RSS like systems (In RSS readers memorable namse were not a concern)
- DAGs revisions carrying IPNS key so that last latest revision or any past revision could be discovered.

> 

Reviewer 5: Also, more advanced mutability systems could be built (e.g. where instead of a single public key, we can have M of N signers required on records). But it all starts from having the UX and process around pinning mutable names and having a performant and efficient mutability system (e.g. Persistent PubSub, like what IPNS over PubSub uses)

### Required resources

##### Effort estimate
<!--T-shirt size rating of the size of the project. If the project might require external collaborators/teams, please note in the roles/skills section below). 
For a team of 3-5 people with the appropriate skills:
- Small, 1-2 weeks
- Medium, 3-5 weeks
- Large, 6-10 weeks
- XLarge, >10 weeks
Describe any choices and uncertainty in this scope estimate. (E.g. Uncertainty in the scope until design work is complete, low uncertainty in execution thereafter.)
-->

> Large. Requires lot of coordinated changes across implementations and collabs. Uncertainty in the scope until concrete actionable plan is in place. Uncertenty in collab buy-in. 

Reviewer 5: True, although just getting things stable + performant in go-ipfs and web browsers would likely go along way while collabs conversations are underway.

Reviewer 3: @Gozala : can we put some high-level estimates (in terms of dev weeks) around each item in the "Brief plan of attack" above?  I think that will help give a bit more clarity on the size of this project.

Author: I'm afraid I'm not well qualified to do this. I've mostly put this proposal together from prior discussions I had with @aschmahmann and @hugomrdias (still in outsider capacity) in an attempt to rally folks who have much better understanding of internal details.

Help trying to scope it out further, would be greatly appreciated.

P.S: Me & @hugomrdias plan no going over this proposal tomorrow to flush it out further.

##### Roles / skills needed
<!--Describe the knowledge/skill-sets and team that are needed for this project (e.g. PM, docs, protocol or library expertise, design expertise, etc.). If this project could be externalized to the community or a team outside PL's direct employment, please note that here.-->

- go-ipfs development
- js-ipfs development

> - cross org coordinator (to get pinning services onboard)

Reviewer 3: Sorry to be slow, but why is it required to get pinning services onboard?  I'm not saying we should skip doing this, I just want to make sure I understand the inter-dependency.  For example, in my head right now I understand that Pinning Services are critical for persistence but I'm less clear on their role in mutability.

Author: That is a valid point and calling out pinning services specifically might be projecting too far of how things would work. 

This was partially influenced by the fact that IPNS (at least in its current form) requires records to be republished otherwise they'll become unsolvable. Implicit  assumption here, is that, unless there are motivated nodes to do this, system will end up impractical. Pinning services just seem well positions to do this given that they are motivated nodes keep content that IPNS will point to available (insert git branch analogy here).

That said we could and probably should consider different strategies to meet reliability goals.

Author: In other words I think we still need someone to help coordinate efforts beyond just leading implementation efforts, to ensure that we do not end up with a system that is directly tied to PL infrastructure. I am assuming this involves getting other teams onboard.

Reviewer 6: +1 on at least thinking about pinning services in the context of IPNS:

- how pinning of /ipns/{libp2p-key-cid} would work?
- how they can help with avoiding IPNS rot 

etc

3 - As high confidence as we can have without doing actual user studies. We know teams that have tried IPNS but found it unrelaible, ultimately rolling out an alternative solution. We also know teams that have evaluated IPFS but chose alternative due to lack of reliable mutability story.

### Project definition
##### Brief plan of attack

<!--Briefly describe the milestones/steps/work needed for this project-->

- Implement third party IPNS republishing (see [ipfs/go-ipfs#4435](https://github.com/ipfs/go-ipfs/issues/4435))
- Collaborate with Pinning Services to make this part of the API.
- Fix IPNS publishing in browsers.
- Enable pubsub by default + IPNS over pubsub.
- Implement & deploy strategic republishing system on PL hosted nodes that will republish popular content and drop stalled one. So that IPNS works even thought pinning services.
- (Nice to have) Implement and rollout delegated publishing to enable updating from multiple devices.
- (Nice to have) Implement and rollout versioning system so that nodes can see full update history.

##### What does done look like?
_What specific deliverables should completed to consider this project done?_

> Publishing IPNS record takes just a few miliseconds. Shutting laptop and loading `https://ipfs.io/ipns/{key}` week later on other device loads the content published as fast as loading corresponding `https://ipfs.io/ipfs/{cid}`.

Reviewer 6: I'd like to apply below change, to make it extra clear the goal here is to solve IPNS rot: 

```suggestion
- Publishing IPNS record takes just a few miliseconds. 
- Shutting laptop and loading `https://ipfs.io/ipns/{key}` week later on other device loads the content published as fast as loading corresponding `https://ipfs.io/ipfs/{cid}`.
- It is possible to pin IPNS address to keep it around and ensure it resolves even when original publisher is gone (AKA solving IPNS rot).
  - Node follows IPNS record updates and keeps re-publishing latest one if original author disappears.
  - Node follows content updates behind the IPNS record, and update content pin every time IPNS record changes. 
  - It is possible to opt-out from this behavior to re-publish IPNS record for `en.wikipedia-on-ipfs.org` without pinning all the data.
```

#####  What does success look like?
_Success means impact. How will we know we did the right thing?_

<!--
Provide success criteria. These might include particular metrics, desired changes in the types of bug reports being filed, desired changes in qualitative user feedback (measured via surveys, etc), etc.
-->
- People and teams are storing IPNS keys in [DNSLink][] records as opposed to CIDs, because that provides a more effective and simpler way to publish updates.
- We see people storing IPNS addresses in ENS to save on blockchain transactions costs.

> - We see new projects leveraging IPNS (when human readable name is not a concern) instead of working around it wit [DNSLink][].

Reviewer 6: This would have to be normalized to account for overall growth, but should be easy to track by processing gateway logs:
```suggestion
- We see new projects leveraging IPNS (when human readable name is not a concern) instead of working around it with [DNSLink][].
- We see more requests for  `/ipns/{libp2p-key}` paths on our public gateway (dweb.link &  ipfs.io).
```

##### Counterpoints &amp; pre-mortem
_Why might this project be lower impact than expected? How could this project fail to complete, or fail to be successful?_

- IPNS has bad reputation, changing it may prove to be a challenge.
- Teams may find that general primitive is not adequate for custom needs.
- Required key management (e.g. key recovery, key compromise) still may prove it impractical.
- Higher level alternatives like textile threads may prove to be better layer of abstraction.

##### Alternatives
_How might this project’s intent be realized in other ways (other than this project proposal)? What other potential solutions can address the same need?_

- Integrating textile threads as core component of the ecosystem.
- Standardizing DNS record updates mechanism and intergating into IPFS.
- Work with Pinning Services to implement stardized DNSLink updating mechanism.

##### Dependencies/prerequisites
<!--List any other projects that are dependencies/prerequisites for this project that is being pitched.-->

- Functioning IPNS in browsers
- Enabling pubsub in go-ipfs

##### Future opportunities
<!--What future projects/opportunities could this project enable?-->

> - [Petname][] system for memorable but non canonical names (kind of how our phones allow us to not remember numbers).

Reviewer 6: I'd like to add dogfooding IPNS in update mechanisms here:
```suggestion
- [Petname][] system for memorable but non canonical names (kind of how our phones allow us to not remember numbers).
- IPNS-based autoupdate mechanisms for our own stack (eg. go-ipfs, ipfs-desktop etc)
```
